backend/src/main/java/fk/prof/backend/http/BackendHttpVerticle.java

```diff
@@ -0,0 +1,177 @@
+        .handler(this::handlePutAssociation);
+
+    return router;
+  }
+
+  @Override
+  public void stop() throws Exception {
+    if(loadReportTimer != null) {
+      vertx.cancelTimer(loadReportTimer);
+    }
+  }
+
+  private void completeStartup(AsyncResult<HttpServer> http, Future<Void> fut) {
+    if (http.succeeded()) {
+      fut.complete();
+    } else {
+      fut.fail(http.cause());
+    }
+  }
+
+  private void handlePostProfile(RoutingContext context) {
+    CompositeByteBufInputStream inputStream = new CompositeByteBufInputStream();
+    RecordedProfileProcessor profileProcessor = new RecordedProfileProcessor(
+        profileWorkService,
+        new SharedMapBasedSingleProcessingOfProfileGate(workIdsInPipeline),
+        config().getJsonObject("parser").getInteger("recordingheader.max.bytes", 1024),
+        config().getJsonObject("parser").getInteger("parser.wse.max.bytes", 1024 * 1024));
+
+    RecordedProfileRequestHandler requestHandler = new RecordedProfileRequestHandler(context, inputStream, profileProcessor);
+    context.request()
+        .handler(requestHandler)
+        .endHandler(v -> {
+          try {
+            if (!context.response().ended()) {
+              //Can safely attempt to close the profile processor here because endHandler is called once the entire body has been read
+              //and example in vertx docs also indicates that this handler will execute once all chunk handlers have completed execution
+              //http://vertx.io/docs/vertx-core/java/#_handling_requests
+              inputStream.close();
+              profileProcessor.close();
+              context.response().end();
+            }
+          } catch (Exception ex) {
+            HttpFailure httpFailure = HttpFailure.failure(ex);
+            HttpHelper.handleFailure(context, httpFailure);
+          }
+        });
+  }
+
+  private void handlePutAssociation(RoutingContext context) {
+    String leaderIPAddress = getLeaderAddressOrAbortRequest(context.response());
+    if (leaderIPAddress != null) {
+      try {
+        //Deserialize to proto message to catch payload related errors early
+        Recorder.ProcessGroup processGroup = Recorder.ProcessGroup.parseFrom(context.getBody().getBytes());
+        makeRequestGetAssociation(leaderIPAddress, processGroup).setHandler(ar -> {
+          if(ar.succeeded()) {
+            context.response().setStatusCode(ar.result().getStatusCode());
+            context.response().end(ar.result().getResponse());
+          } else {
+            HttpFailure httpFailure = HttpFailure.failure(ar.cause());
+            HttpHelper.handleFailure(context, httpFailure);
+          }
+        });
+      } catch (Exception ex) {
+        HttpFailure httpFailure = HttpFailure.failure(ex);
+        HttpHelper.handleFailure(context, httpFailure);
+      }
+    }
+  }
+
+  private String getLeaderAddressOrAbortRequest(HttpServerResponse response) {
```

Comment:
This is handy wrapper to call before handling any request which needs to be proxied. It encapsulates the logic of verifying if self is leader, if there is no leader and responds accordingly